### PR TITLE
Reduced code duplication in PMFT and RDF

### PR DESCRIPTION
**Pull request by jinsooihm**
## Description

Added `accumulateGeneral`, `resetGeneral` and other various convenience functions in `NdHistogram` class to capture repeated logic in other `PMFT` subclasses and `RDF`.
One caveat is that the structure of `NdHistogram` is very specific to `PMFT` and `RDF` that it would not really help someone who is trying to make a new histogram class. 
Also, it might actually just make more sense to make it as a friend class that just provides useful functionality instead of a parent.

## Motivation and Context

Modules that are histogram based have a lot of repeated logic.
`CorrelationFunction` can be included in the framework, if we make `NdHistogram` to be templated, and perhaps make #295 more generalized to use in `accumulate`. 
Reduction is sufficiently different that it seems it should be implemented separately.

Possible ideas.
1. CorrelationFunction has two different arrays that it computes in parallel, while all others have one array. We can make the initialization check for each threads to be a member function or lambda. (Does that work correctly with `tbb::enumerable_thread_specific.local()`?)
2. Make #295 be more general, i.e. able to do some checks between `parallel_for` and actual loop between neighbors and use that in `accumulate`. But this might make things just a bit more convoluted.
    

## How Has This Been Tested?

Passes all the current tests.

## Types of changes

- [x] New feature (non-breaking change which adds or improves functionality)

## Checklist:

- [ ] My code follows the code style of this project. 
   (Not `clang-format`ed yet)
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cpp/pmft/PMFTXY2D.cc [hunk 1]
```diff
@@ -41,177 +41,91 @@ PMFTXY2D::PMFTXY2D(float x_max, float y_max, unsigned int n_x, unsigned int n_y)
...
+    Index2D b_i = Index2D(m_n_x, m_n_y);
+
+    accumulateGeneral(box, n_ref, nlist, n_p, m_n_x * m_n_y, [=](size_t i, size_t j) {
+        vec3<float> ref = ref_points[i];
+        vec3<float> delta = this->m_box.wrap(points[j] - ref);
+        float rsq = dot(delta, delta);
+
+        // check that the particle is not checking itself
+        // 1e-6 is an arbitrary value that could be set differently if needed
+        if (rsq < 1e-6)
```
Comment (bdice) [inline cpp/pmft/PMFTXY2D.cc:101]: Oh gosh. This needs to be handled differently with `exclude_ii`-type logic.

File: cpp/pmft/PMFTXYT.cc [hunk 2]
```diff
@@ -47,189 +46,93 @@ PMFTXYT::PMFTXYT(float x_max, float y_max, unsigned int n_x, unsigned int n_y, u
...
-        [=] (const blocked_range<size_t>& r)
-            {
-            for (size_t i = r.begin(); i != r.end(); i++)
-                {
-                m_pcf_array.get()[i] = (float)m_bin_counts.get()[i] * norm_factor * inv_jacobian * inv_num_dens;
-                }
-            });
-    }
+{
+    float jocob_factor = (float) 1.0 / m_jacobian;
```
Comment (bdice) [inline cpp/pmft/PMFTXYT.cc]: Rename this: `jacobian_factor`

File: cpp/util/NdHistogram.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+#include <memory>
+#include <tbb/tbb.h>
+
+#include "Box.h"
+#include "NeighborList.h"
+
+namespace freud { namespace util {
+
+template<typename T>
+std::shared_ptr<T> returnEmptyArray(unsigned int size)
```
Comment (bdice) [inline cpp/util/NdHistogram.h]: Rename to `makeEmptyArray`?

File: cpp/util/NdHistogram.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    protected:
+        box::Box m_box;
+        unsigned int m_frame_counter; //!< Number of frames calculated
+        unsigned int m_n_ref;         //!< The number of reference points
+        unsigned int m_n_p;           //!< The number of points
+        bool m_reduce;                //!< Whether or not the PCF needs to be reduced
+
+        std::shared_ptr<float> m_pcf_array;         //!< Array of PCF/RDF computed
```
Comment (bdice) [inline cpp/util/NdHistogram.h]: Specify that PCF stands for Pair Correlation Function in the comment. `//!< Array of computed pair correlation function`

File: cpp/density/RDF.cc [hunk 5]
```diff
@@ -28,262 +27,187 @@ RDF::RDF(float rmax, float dr, float rmin)
...
 //! CumulativeCount class to perform a parallel reduce to get the cumulative count for each histogram bin
 class CumulativeCount
+{
+private:
+    float m_sum;
+    float* m_N_r_array;
+    float* m_avg_counts;
+
+public:
+    CumulativeCount(float* N_r_array, float* avg_counts)
```
Comment (bdice) [inline cpp/density/RDF.cc]: This entire `CumulativeCount` class is a holdover from TBB before lambdas. This should definitely be rewritten.
https://www.threadingbuildingblocks.org/docs/help/reference/algorithms/parallel_scan_func.html

**Review by bdice — COMMENTED** [review]
Some comments below. More review is needed.

File: cpp/pmft/PMFTXY2D.cc (hunk 1 above)
Comment (jinsooihm) [reply cpp/pmft/PMFTXY2D.cc:101]: Is it guaranteed that if the indices of two points are same, then the two points are same points?

File: cpp/density/RDF.cc (hunk 5 above)
Comment (jinsooihm) [reply cpp/density/RDF.cc]: For some reason, it seems that tbb on Ubuntu cannot find the template for
```
template<typename Range, typename Value, typename Scan, typename Combine>
Value parallel_scan( const Range& range, const Value& identity,
                     const Scan& scan, const Combine& combine
                     [, partitioner] );
```
The implementation using this version fails to compile on Ubuntu containers on circleci, but compiles on my computer (Mac) and arch container.

Comment (jinsooihm) [reply cpp/density/RDF.cc]: The reason is that this `parallel_scan` was added for the 2018 update 1 for `tbb` but the default on Ubuntu is using the 2017 version.

File: benchmarks/benchmark_PMFT_PMFTR12.py [hunk 6]
```diff
@@ -24,6 +24,7 @@ def bench_setup(self, N):
 
     def bench_run(self, N):
         self.pmft.compute(self.box, self.points, self.orientations)
+        self.pmft.bin_counts
```
Comment (vyasr) [inline benchmarks/benchmark_PMFT_PMFTR12.py:27]: We can remove this, right? The computation isn't lazy so we don't need to benchmark the access

Comment (vyasr) [reply benchmarks/benchmark_PMFT_PMFTR12.py:27]: Remove below as well, unless I'm missing something.

File: cpp/density/RDF.cc [hunk 7]
```diff
@@ -28,262 +27,192 @@ RDF::RDF(float rmax, float dr, float rmin)
...
-            m_sum = b.m_sum;
-            }
-    };
-
+        m_sum = temp;
+    }
+    CumulativeCount( CumulativeCount& b, split )
+        : m_sum(0), m_N_r_array(b.m_N_r_array), m_avg_counts(b.m_avg_counts)
+    {
+    }
```
Comment (vyasr) [inline cpp/density/RDF.cc:92]: Does our clang style suggest always splitting this line? Just wondering.

**Review by vyasr — CHANGES_REQUESTED** [review]
@jinsoo960 could you fix merge conflicts and merge in master, then I'll review again? I'm having a bit of a hard time separating out which lines are actually changed and which are due to clang-formatting, and I'm worried that I'll miss something important as a result.

File: benchmarks/benchmark_PMFT_PMFTR12.py (hunk 6 above)
Comment (jinsooihm) [reply benchmarks/benchmark_PMFT_PMFTR12.py:27]: I believe `reduce` does not get called in `compute`, but only when you try to access `bin_counts`.

File: cpp/density/RDF.cc (hunk 7 above)
Comment (jinsooihm) [reply cpp/density/RDF.cc:92]: Seems like it does not split it when I tried right now.

File: cpp/pmft/PMFTXY2D.cc (hunk 1 above)
Comment (bdice) [reply cpp/pmft/PMFTXY2D.cc:101]: We are fixing this in #328.

**Comment by bdice** [comment]
@jinsoo960 wants to remove 4 lines of code before merging this.

edit: nevermind.

**Review by bdice — APPROVED** [review]

**Pull request merged**